### PR TITLE
Add function factorial

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add function factorial, as in Impala or Spark. 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: docs/en/sql-reference/functions/math-functions.md [hunk 1]
```diff
@@ -549,3 +549,33 @@ Result:
 │ 3.141592653589793 │
 └───────────────────┘
 ```
+
+
+## factorial(n)
+
+Computes the factorial of an integer value. It works with any native integer type including UInt(8|16|32|64) and Int(8|16|32|64). The return type is Int64.
+
+The factorial of 0 is 1. Likewise, the factorial() function returns 1 for any negative value. The maximum positive value for the input argument is 20; a value of 21 or greater overflows the range for Int64 and will causes an throw exception.
```
Comment (kitaisreal) [inline docs/en/sql-reference/functions/math-functions.md:558]: Consider to rename `will causes an throw` into `will throw exception`, `will cause exception throw`.

File: src/Functions/factorial.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+{
+    static bool has() { return true; }
+
+    static IFunction::Monotonicity get(const Field & /*left*/, const Field & /*right*/)
+    {
+        return {
+            .is_monotonic = true,
+            .is_positive = true,
+            .is_always_monotonic = true,
+            .is_strict = false,
```
Comment (kitaisreal) [inline src/Functions/factorial.cpp:83]: I suspect `is_strict` can be `true` ?

File: src/Functions/factorial.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+REGISTER_FUNCTION(Factorial)
+{
+    factory.registerFunction<FunctionFactorial>(
+        {
+            R"(
+Computes the factorial of an integer value. It works with any native integer type. The return type is Int64.
+
+The factorial of 0 is 1. Likewise, the factorial() function returns 1 for any negative value. The maximum positive value for the input argument is 20; a value of 21 or greater overflows the range for Int64 and will causes an throw exception.
```
Comment (kitaisreal) [inline src/Functions/factorial.cpp:96]: Same comments as above for documentation.

File: docs/en/sql-reference/functions/math-functions.md (hunk 1 above)
Comment (kitaisreal) [reply docs/en/sql-reference/functions/math-functions.md:558]: I also `is 20; a` maybe better to rename into `is 20, a`.

File: tests/queries/0_stateless/00087_math_functions.sql [hunk 4]
```diff
@@ -137,3 +137,10 @@ select erf(10) = 1;
 select erfc(0) = 1;
 select erfc(-10) = 2;
 select erfc(28) = 0;
+
+select factorial(-1) = 1;
```
Comment (kitaisreal) [inline tests/queries/0_stateless/00087_math_functions.sql:141]: Better to create new test, for us it will be easier to track if test becomes flacky.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Looks good. Added a couple of small comments.

**Comment by taiyang-li** [comment]
Failed tests not related to this PR.

**Review by kitaisreal — APPROVED** [review]

File: src/Functions/factorial.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+template <typename A>
+struct FactorialImpl
+{
+    using ResultType = Int64;
```
Comment (kitaisreal) [inline src/Functions/factorial.cpp:18]: Why do we return `Int64` intead of `UInt64`, if function only returns values greater than 1 ?

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Functions/factorial.cpp (hunk 5 above)
Comment (taiyang-li) [reply src/Functions/factorial.cpp:18]: Because I wanted to keep result type same with Impala or Spark.

Comment (taiyang-li) [reply src/Functions/factorial.cpp:18]: You are right. I will modify it.

Comment (taiyang-li) [reply src/Functions/factorial.cpp]: @kitaisreal  done

File: docs/en/sql-reference/functions/math-functions.md [hunk 6]
```diff
@@ -549,3 +549,33 @@ Result:
 │ 3.141592653589793 │
 └───────────────────┘
 ```
+
+
+## factorial(n)
+
+Computes the factorial of an integer value. It works with any native integer type including UInt(8|16|32|64) and Int(8|16|32|64). The return type is UInt64.
+
+The factorial of 0 is 1. Likewise, the factorial() function returns 1 for any negative value. The maximum positive value for the input argument is 20, a value of 21 or greater overflows the range for Int64 and will cause exception throw.
```
Comment (kitaisreal) [inline docs/en/sql-reference/functions/math-functions.md:558]: `for Int64` need to be renamed into `for UInt64`

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Looks good to me, need to fix documentation and merge

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**